### PR TITLE
compactor: change GatherIndexIssueStats validation logic on sorted labels

### PR DESCRIPTION
**Pull request by sgmitchell**
Deliberately exclude the first element when checking that a label set is in sorted order.

The first element is always `__name__` for prometheus and while `_` is lexicographically before all lowercase letters, it does not precede upper case letters. This made compaction fail on valid metrics such as `up{Z="one", a="two"}` since the expected order was `[Z, __name__, a]`.

This change brings the validation more in line with what prometheus is using in [promparse](https://github.com/prometheus/prometheus/blob/9f903fb3f7841f55d1b634d5fb02986a8bfc5e0c/pkg/textparse/promparse.go#L230).

An additional check may also be needed to make sure that the first value is always `__name__`

## Changes

* Add test cases for `GatherIndexIssueStats`
* Change sorted label validation to skip the first element

## Verification

Here's a simple config that generates data that will crash the compactor. I validated that my fix no longer crashed the compactor on a similar dataset.

Use the following scrape config in prometheus to generate a metrics `up{Z="one", a="two", instance="localhost:9090",job="prometheus"}`
```
scrape_configs:
  - job_name: 'prometheus'
    static_configs:
    - targets:
      - 'localhost:9090'
      labels:
        Z: one
        a: two
```

Start the compactor running the `improbable/thanos:v0.3.0` image and read the bucket that the thanos sidecar is writing. When the container encounters this timeseries, it will crash with the the following log line

`level=error ts=... caller=main.go:181 msg="running command failed" err="error executing compaction: compaction failed: compaction: gather index issues for block /var/thanos/store/compact/0@{monitor=\"prometheus\",replica=\"prometheus-0\"}/01D36C79GDX5DVXZ1PAD5ZT65M: out-of-order label set {__name__=\"up\",Z=\"one\",a=\"two\",instance=\"localhost:9090\",job=\"prometheus\"} for series 2739605"
`

### CODE REVIEW

**Comment by sgmitchell** [comment]
> upper case label names are quite wrong

Agreed, although I didn't know about the perf hit aspect of it. The metrics that tickled this bug were changed as soon as we found them but unfortunately, we can't rewrite history, so the compactor needs to be flexible enough to deal with (and eventually erase) our bad decisions.

Related to this: I'd be interested if you can you point me to any discussions or pages that talk about this perf hit?
I know that almost every metric I see conforms to lower and snake case, but when I looked at [the best practices](https://prometheus.io/docs/practices/naming/), there was no explicit call out.

Thanks for the quick reply

File: pkg/block/index.go [hunk 1]
```diff
@@ -345,13 +345,12 @@ func GatherIndexIssueStats(logger log.Logger, fn string, minTime int64, maxTime
...
 		}
-		l0 := lset[0]
-		for _, l := range lset[1:] {
-			if l.Name <= l0.Name {
-				return stats, errors.Errorf("out-of-order label set %s for series %d", lset, id)
-			}
-			l0 = l
+
+		// Ensure all but the first value (__name__) are sorted
+		if !sort.SliceIsSorted(lset[1:], func(i, j int) bool { return lset[i+1].Name <= lset[j+1].Name }) {
```
Comment (bwplotka) [inline pkg/block/index.go]: Let's actually check if __name__ is first maybe?

File: pkg/block/index.go [hunk 2]
```diff
@@ -345,13 +345,12 @@ func GatherIndexIssueStats(logger log.Logger, fn string, minTime int64, maxTime
...
 			return stats, errors.Errorf("series %v out of order; previous %v", lset, lastLset)
 		}
-		l0 := lset[0]
-		for _, l := range lset[1:] {
-			if l.Name <= l0.Name {
-				return stats, errors.Errorf("out-of-order label set %s for series %d", lset, id)
-			}
-			l0 = l
+
+		// Ensure all but the first value (__name__) are sorted
```
Comment (bwplotka) [inline pkg/block/index.go]: missing trailing petiof

File: pkg/block/index.go (hunk 1 above)
Comment (bwplotka) [reply pkg/block/index.go]: because I think that's the whole point.. if the labels are ordered __name__ is between uppercase and lowercase already.

**Comment by sgmitchell** [comment]
I initially omitted the check for `lset[0] == "__name__"` because I only found a few occurrences of `__name__` in [prometheus/tsdb](https://github.com/prometheus/tsdb) and none of them indicated to me that it **had** to be the first label. The only place I see that invariant is in [prometheus/prometheus](https://github.com/prometheus/prometheus). 
Since the `blocks` package seemed like it was much more closely aligned with the tsdb code, I was hesitant to add that logic.

[Issue #32 in prometheus/tsdb](https://github.com/prometheus/tsdb/issues/32) was the best guidance I found on how this should be implemented and after reading your comment, I've added [another test case]() that checks bad (at least in my eyes) labelsets correctly throw an error. Unfortunately, the old code happily accepts these label sets while my new changes result in an error (the expected output). So now I'm not 100% sure this won't break anything 🤷‍♂️.

**Comment by sgmitchell** [comment]
Yuck. Ensuring `__name__` is first [failed e2e testing](https://circleci.com/gh/improbable-eng/thanos/2241) where the labelset was `{a="1"}`. 

So I guess the question becomes: when is `__name__` required in the label set for a tsdb? If it's always, then we need to change the test data to reflect that. If it's anything less than always, I'm not sure what the best and safest path forward is.

**Comment by bwplotka** [comment]
Let's focus on what we want to achieve here. Are you blocked? Do you use upper case label names? (:

File: pkg/block/index.go [hunk 3]
```diff
@@ -345,13 +347,17 @@ func GatherIndexIssueStats(logger log.Logger, fn string, minTime int64, maxTime
...
-			}
-			l0 = l
+
+		// Ensure that __name__ comes first.
+		if lset[0].Name != model.MetricNameLabel {
+			return stats, errors.Errorf("label set %s does not start with %s for series %d", lset, model.MetricNameLabel, id)
 		}
+
+		// Ensure all but the first value (__name__) are sorted.
+		if !sort.SliceIsSorted(lset[1:], func(i, j int) bool { return lset[i+1].Name <= lset[j+1].Name }) {
```
Comment (GiedriusS) [inline pkg/block/index.go]: It's a bit different and I might raise a different pull request but what if we would relax this check a bit and switch it to `<` here? Essentially, it seems like Prometheus accepts duplicate labels and if that happens (happened in our case) then compactor doesn't work anymore due to this check. Case in point:

```
{"caller":"main.go:181","err":"error executing compaction: compaction failed: compaction: gather index issues for block /data/compact/0@{monitor=\"monitor\",replica=\"repl\"}/01D34EDQMSQ29RHAC47XGKHGC7: out-of-order label set {__name__=\"foo\",exported_job=\"vv\",host=\"172_16_226_56\",host=\"172_16_226_56\",region=\"lt\",subtask_index=\"5\",task_attempt_id=\"32e4b047bb768583ff57c709be3b1046\",task_attempt_num=\"8\",task_id=\"688c028a219ff3372f3eecb0ee5811f9\",task_name=\"Source:_foo\",tenant=\"abc\",tier=\"cooltier\",tm_id=\"53b2ed987b08f427dec4ee1465df91fa\"} for series 2594231","level":"error","msg":"running command failed","ts":"2019-02-11T13:30:33.901722306Z"}
```

**Comment by sgmitchell** [comment]
> Let's focus on what we want to achieve here. Are you blocked? Do you use upper case label names? (:

I'm dogfooding my fork and it's running great so far. The metrics with capital labels are gone but unfortunately, I'll still need to use my fork until those troublesome metrics age out of the retention window (which by design will be a while).
This PR _is_ probably blocked until there is a determination how to reconcile the 2 differences.

I think the ultimate goal should be that compactor validation should be as close to prometheus' validation but no more strict than prometheus. This PR specifically should just cover the edge case of labels that start with a capital letter.

A temporary fix until we can agree on the right impl here, might be completing #711.

File: pkg/block/index.go (hunk 3 above)
Comment (GiedriusS) [reply pkg/block/index.go]: Moved to https://github.com/improbable-eng/thanos/pull/848.

**Comment by bwplotka** [comment]
Ok, so with https://github.com/improbable-eng/thanos/pull/848 being merged, can we rebase this / or decide to close? (:  There is definitely more work on this if we want to productionize that (:

**Comment by sgmitchell** [comment]
I'm ok with closing this.
#848 won't fix this issue but unfortunately I don't currently have the bandwidth to contribute more.
I've corrected my metrics with uppercase labels to be lowercase, which has gotten past this problem but it seems like there needs to be some work to ensuring thanos and prometheus don't differ in their checks

**Comment by ipstatic** [comment]
I just ran into this bug. @sgmitchell does this PR still fix the issue?

**Comment by bwplotka** [comment]
Yea, some idea is implement this:
https://github.com/improbable-eng/thanos/issues/415 to allow back in time
fixes.

śr., 13 mar 2019 o 17:10 Jarod Watkins <notifications@github.com>
napisał(a):

**Comment by GiedriusS** [comment]
I think this is still valid. As I understand, `__name__` should always be first because for metric names and label names the same PromQL restrictions apply as per https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md. Then, from this follows that first label should always be `__name__`, and then the rest of them should be sorted in a lexicographical order. TSDB doesn't care about `__name__ ` but Prometheus always adds it :P So, in my view, the tests should be fixed to include `__name__` in the test data, and then we should merge this.

**Comment by ipstatic** [comment]
I just built this branch (after rebasing off master) and it does not solve my issue (some labels that are all uppercase). You can see my update here, maybe I did the merge wrong?

https://github.com/ipstatic/thanos/commit/0dd3cb045277c27ddd7cfe0c4e412755509460f2

**Comment by jjneely** [comment]
Ok, rebased myself to double check @ipstatic's work.  

https://github.com/jjneely/thanos/tree/GatherIndexIssueStats-label-sorting

I updated my compact docker image and re-ran, and I get an error that `__name__` isn't the first label in the set.  From the error message, that looks totally correct.  

Can we at least skip compaction of bad blocks?  Or re-order the list correctly here?

**Comment by GiedriusS** [comment]
Ok, looked a bit more into this and it seems like in the Prometheus code there is some single package for all things related to labels: https://github.com/prometheus/prometheus/blob/9f903fb3f7841f55d1b634d5fb02986a8bfc5e0c/pkg/labels/labels.go. As you can see, everywhere there the label set is being sorted before being returned to the users of that package. In here as mentioned in the original issue: https://github.com/prometheus/prometheus/blob/9f903fb3f7841f55d1b634d5fb02986a8bfc5e0c/pkg/textparse/promparse.go#L230 the `__name__` label is put at the front which breaks the invariant if any of the other labels use uppercase letters. So it seems like it is a bug there in `promparse`... thoughts?

**Comment by GiedriusS** [comment]
@jjneely indeed this seems like a Prometheus bug. I added another test to the parser in Prometheus and ran the tests with: `wind_speed{A="2",c="3"} 12345`.

Result:

```
--- FAIL: TestPromParse (0.00s)
    /Users/statkg/go/src/github.com/prometheus/prometheus/pkg/textparse/require.go:157: 
        	Error Trace:	promparse_test.go:193
        	Error:      	Not equal: 
        	            	expected: labels.Labels{labels.Label{Name:"A", Value:"2"}, labels.Label{Name:"__name__", Value:"wind_speed"}, labels.Label{Name:"c", Value:"3"}}
        	            	actual  : labels.Labels{labels.Label{Name:"__name__", Value:"wind_speed"}, labels.Label{Name:"A", Value:"2"}, labels.Label{Name:"c", Value:"3"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(labels.Labels) (len=3) {A="2", __name__="wind_speed", c="3"}
        	            	+(labels.Labels) (len=3) {__name__="wind_speed", A="2", c="3"}
        	            	 
        	Test:       	TestPromParse
```
So yes, it seems that Prometheus parser itself isn't following the invariants laid out in the labels package. Needs fixing there, IMHO.

**Comment by GiedriusS** [comment]
@jjneely https://github.com/prometheus/prometheus/pull/5372 :)

**Comment by jjneely** [comment]
Yay!  So labels values should always be in sorted order with no special treatment of `__name__`.  So the bug/corruption I'm suffering from is that I have postings that have label sets ordered like this:

```
{__name__=\"DISCOVERED_ALERTS\",DATA_CENTER=\"dal09\",alertname=\"PuppetAgentOpenFdsHigh\",alertstate=\"healthy\",description=\"{{$labels.instance}} of job {{$labels.job}} has been elevated (over 80% of hard limit) for more than 5 minutes.\",instance=\"mesos-10-153-2-69-x2.prod.dal09.example.com:31255\",job=\"http_status-page_prod_site-svc-status-page_dal09\",monitor=\"dal09/sre/prod/prometheus\",runbook=\"https://wiki.example.com/display/siteops/number+of+open+files\",severity=\"page\",summary=\"{{$labels.instance}} puppet-agent open_fds high\"}
```

Which was the original error I was facing that lead me here.

**Comment by GiedriusS** [comment]
Yes. So now we should add an extra flag to Thanos Compact to make it ignore this error, and print information about this in case it happens. We should explicitly note that all versions before and including Prometheus 2.8.0 are affected by this. Care to make a patch?

**Comment by GiedriusS** [comment]
Closing this because this has been outlined already in https://github.com/improbable-eng/thanos/issues/888. Will add the relevant info there.

**Comment by jjneely** [comment]
Yes.  I'll be glad to work on a PR for this.  Thanks a bunch for the help, makes contributing easy.

**Comment by ashurov-os33** [comment]
Is it possible to disable sorting?

**Comment by GiedriusS** [comment]
No, probably because it permits Prometheus to use binary search. What problem are you trying to solve?

**Comment by ashurov-os33** [comment]
I use HTTP URL parameters. (target, user, pass). Because of the sorting, I have the wrong request url.
I have http://server/metrics?pass=123&target=1.1.1.1&user=foo
I need http://server/metrics?target=1.1.1.1&user=foo&pass=123

**Comment by GiedriusS** [comment]
I'm sorry but I think that your issue is completely unrelated to this one and even the Thanos project 😄 pleask ask in https://github.com/prometheus/prometheus about this. But most likely you will be told that your application must not depend on such minute details because IMHO it would be hard to implement any kind of ability to specify an arbitrary ordering of parameters when scraping.

**Pull request closed without merging**